### PR TITLE
Tests for CollectionStream

### PR DESCRIPTION
**Pull request by dvdplm**
Based on https://github.com/libp2p/rust-libp2p/pull/584.

### CODE REVIEW

File: core/src/nodes/collection.rs [hunk 1]
```diff
@@ -464,3 +464,382 @@ impl<'a, TInEvent> PeerMut<'a, TInEvent> {
...
+    use nodes::NodeHandlerEvent;
+    use std::sync::Arc;
+    use parking_lot::Mutex;
+
+    type TestCollectionStream = CollectionStream<InEvent, OutEvent, Handler>;
+
+    #[test]
+    fn can_create_new() {
+        let cs = TestCollectionStream::new();
+        assert_matches!(cs.inner, HandledNodesTasks{..});
```
Comment (tomaka) [inline core/src/nodes/collection.rs]: We shouldn't test the internals.

File: core/src/nodes/collection.rs [hunk 2]
```diff
@@ -464,3 +464,382 @@ impl<'a, TInEvent> PeerMut<'a, TInEvent> {
...
+        let mut cs = TestCollectionStream::new();
+        let fut = future::empty();
+        let valid_reach_id = cs.add_reach_attempt(fut, Handler::default());
+
+        // Obtain a TaskId we can use to build a ReachAttemptId that is not in
+        // the list of tasks in the CollectionStream we're testing. Note that
+        // TaskIds with the same number are all considered equal, so TaskId(0)
+        // from one HandledNodesTask is == to TaskId(0) from another
+        // HandledNodesTask.
+        let invalid_reach_id = {
```
Comment (tomaka) [inline core/src/nodes/collection.rs]: I don't think that's a nice way of doing it. You're relying on the fact that the task ID is linear, which is not necessarily true.
A better test, IMO, is to interrupt the same task twice and check that it is an error the second time.

Comment (dvdplm) [reply core/src/nodes/collection.rs]: The original idea with this test was this: prove that I can't misuse a `TaskId` from reach attempts belonging to different `CollectionStream`s, to interrupt one using a `TaskId` from the other. That's actually allowed though (and not really a concern we care about I'd say).

Interrupting twice is a slightly different thing, but exercise the same code path so I'll add that. 

And remove this test.

File: core/src/nodes/collection.rs (hunk 1 above)
Comment (dvdplm) [reply core/src/nodes/collection.rs]: removing

**Pull request merged**